nodejs/awsx/ec2/vpcTopology.ts

```diff
@@ -147,14 +359,83 @@ ${lastAllocatedIpAddress} > ${lastVpcIpAddress}`);
             return `${vpcName}-${subnetName}`;
         }
     }
+
+    private createIpv6CidrBlock(
+            assignIpv6AddressOnCreation: pulumi.Input<boolean>,
+            index: number): pulumi.Output<string> {
+
+        const result = pulumi.all([this.ipv6CidrBlock, assignIpv6AddressOnCreation])
+                     .apply(([vpcIpv6CidrBlock, assignIpv6AddressOnCreation]) => {
+                if (!assignIpv6AddressOnCreation) {
+                    return undefined;
+                }
+
+                if (!vpcIpv6CidrBlock) {
+                    throw new pulumi.ResourceError(
+"Must set [assignGeneratedIpv6CidrBlock] to true on [Vpc] in order to assign ipv6 address to subnet.", this.resource);
+                }
+
+                // Should be of the form: 2600:1f16:110:2600::/56
+                const colonColonIndex = vpcIpv6CidrBlock.indexOf("::");
+                if (colonColonIndex < 0 ||
+                    vpcIpv6CidrBlock.substr(colonColonIndex) !== "::/56") {
+
+                    throw new pulumi.ResourceError(`Vpc ipv6 cidr block was not in an expected form: ${vpcIpv6CidrBlock}`, this.resource);
+                }
+
+                const header = vpcIpv6CidrBlock.substr(0, colonColonIndex);
+                if (!header.endsWith("00")) {
+                    throw new pulumi.ResourceError(`Vpc ipv6 cidr block was not in an expected form: ${vpcIpv6CidrBlock}`, this.resource);
+                }
+
+                // trim off the 00, and then add 00, 01, 02, 03, etc.
+                const prefix = header.substr(0, header.length - 2);
+                return prefix + index.toString().padStart(2, "0") + "::/64";
+             });
+
+        return <pulumi.Output<string>>result;
+    }
+}
+
+function shouldCreateNatGateways(
+        numberOfNatGateways: number, publicSubnets: SubnetDescription[], privateSubnets: SubnetDescription[]) {
+    // To make natgateways:
+    // 1. we have to have at least been asked to make some nat gateways.
+    // 2. we need public subnets to actually place the nat gateways in.
+    // 3. we need private subnets that will actually be connected to the nat gateways.
+    return  numberOfNatGateways > 0 && publicSubnets.length > 0 && privateSubnets.length > 0;
 }
 
-interface SubnetDescription {
+/** @internal */
+export interface VpcTopologyDescription {
+    subnets: SubnetDescription[];
+    natGateways: NatGatewayDescription[];
+    natRoutes: NatRouteDescription[];
+}
+
+/** @internal */
+export interface SubnetDescription {
     type: x.ec2.VpcSubnetType;
     subnetName: string;
-    availabilityZone: number;
-    cidrBlock: string;
-    tags?: pulumi.Input<aws.Tags>;
-    mapPublicIpOnLaunch?: pulumi.Input<boolean>;
-    assignIpv6AddressOnCreation?: pulumi.Input<boolean>;
+    args: x.ec2.SubnetArgs;
```

Comment:
simpler design.  VpcTopology creates the info about a subnet, which is just it's type, it's name, and then then args to construct an awsx.ec2.Subnet.